### PR TITLE
Adsk Contrib - Improve FileTransform interpolation

### PR DESCRIPTION
**Pull request by BernardLefebvre**
- Changes the FileTransform default interpolation from unknown to default so that FileTransform can be used without setting the interpolation type. 
- Refactor how FileFormats are using the interpolation.  If a specified interpolation can not be used, log a warning.  See comments in FileFormat3DL.cpp and tests in FileFormatSpi3D_tests.cpp.
- CTF/CLF now uses the interpolation provided for LUTs if they do not specify an interpolation within the file itself.
- Fix issue where two FileTransforms pointing to the same file but using different interpolation could be problematic.
- Add interpolation check to FileTransform validation.
- This fixes issue #1121 (test has been added to FileTransformTest.py).
- Add a getTransformType method to the Transform classes to ease the instance type check.

Signed-off-by: Bernard Lefebvre <bernard.lefebvre@autodesk.com>

### CODE REVIEW

**Comment by BernardLefebvre** [comment]
I checked OCIO v1 and did confirm that LUT file are throwing if not interpolation is porvided.
For LUT1D, exception is "Cannot apply Lut1DOp, unspecified interpolation." from https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/RB-1.1/src/core/Lut1DOp.cpp line 562.

The following test (using local files) does confirm this.
`
OIIO_ADD_TEST(FileTransform, interpolation)
{
    std::string SIMPLE_PROFILE =
        "ocio_profile_version: 1\n"
        "resource_path: luts\n"
        "strictparsing: false\n"
        "luma: [0.2126, 0.7152, 0.0722]\n"
        "roles:\n"
        "  default: raw\n"
        "colorspaces:\n"
        "  - !<ColorSpace>\n"
        "      name: raw\n"
        "      family: raw\n"
        "      equalitygroup: \n"
        "      bitdepth: 32f\n"
        "      description: |\n"
        "        A raw color space. Conversions to and from this space are no-ops.\n"
        "      isdata: true\n"
        "      allocation: uniform\n"
        "  - !<ColorSpace>\n"
        "      name: lnh\n"
        "      family: ln\n"
        "      equalitygroup: \n"
        "      bitdepth: 16f\n"
        "      isdata: false\n"
        "      allocation: lg2\n"
        "  - !<ColorSpace>\n"
        "      name: lut_file\n"
        "      to_reference: !<FileTransform>  {src: D:/OCIO/data/lut.spi1d}\n"
        "  - !<ColorSpace>\n"
        "      name: matrix_file\n"
        "      to_reference: !<FileTransform>  {src: D:/OCIO/data/matrix.spimtx}\n"
        "\n";

    std::istringstream is;
    is.str(SIMPLE_PROFILE);
    OCIO::ConstConfigRcPtr config;
    OIIO_CHECK_NO_THROW(config = OCIO::Config::CreateFromStream(is));

    auto p1 = config->getProcessor("lnh", "matrix_file");
    OIIO_CHECK_ASSERT(p1);
    OIIO_CHECK_THROW(config->getProcessor("lnh", "lut_file"), OCIO::Exception);
}
`

**Comment by doug-walker** [comment]
@Shootfast , @michdolan , this completes our action item from the TSC meeting and confirms our statement that OCIO v1 throws if no interpolation is provided for LUT files.  Perhaps you were thinking of some other file type such Matrix, which does not throw (since interpolation is not relevant for that transform).  In any case, this PR now avoids the throw in all cases by using default interpolation if it is unspecified.

File: src/OpenColorIO/fileformats/FileFormatUtils.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#include "ops/lut1d/Lut1DOpData.h"
+#include "ops/lut3d/Lut3DOpData.h"
+
+namespace OCIO_NAMESPACE
+{
+Lut1DOpDataRcPtr HandleLUT1D(const Lut1DOpDataRcPtr & fileLut1D,
+                             Interpolation fileInterp,
+                             bool & fileInterpUsed);
+
+Lut3DOpDataRcPtr HandleLUT3D(const Lut3DOpDataRcPtr & fileLut1D,
```
[the remark this answers is not in the document]
Comment (BernardLefebvre) [reply src/OpenColorIO/fileformats/FileFormatUtils.h]: Fixed

File: tests/cpu/fileformats/FileFormatCTF_tests.cpp [hunk 2]
```diff
@@ -7718,6 +7773,70 @@ OCIO_ADD_TEST(CTFTransform, lut3d_inverse_ctf)
...
+    lut->setInterpolation(OCIO::INTERP_LINEAR);
+    processor = config->getProcessor(lut);
+    OCIO_CHECK_NO_THROW(processor->write(OCIO::FILEFORMAT_CTF, outputTransform));
+
+    {
+        const std::string result = outputTransform.str();
+        const std::string expected = R"(<LUT3D inBitDepth="32f" outBitDepth="32f" interpolation="trilinear">)";
+        OCIO_CHECK_ASSERT(result.find(expected) != std::string::npos);
+    }
+
```
[the remark this answers is not in the document]
Comment (BernardLefebvre) [reply tests/cpu/fileformats/FileFormatCTF_tests.cpp:7821]: Fixed

**Comment by hodoulp** [comment]
As usual this is a notice that we will merge this on Monday, if there are no more discussions on this pull request.

**Review by doug-walker — APPROVED** [review]

**Review by hodoulp — APPROVED** [review]

**Pull request merged**